### DISCUSSION TITLE
chapter_divide_and_conquer/divide_and_conquer/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#610] [Announcements]
# chapter_divide_and_conquer/divide_and_conquer/

动画图解、一键运行的数据结构与算法教程

https://www.hello-algo.com/chapter_divide_and_conquer/divide_and_conquer/

**Comment by 1104797544** [comment]
1.   操作数量优化¶
以“冒泡排序”为例，其处理一个长度为 \(n\) 的数组需要 \(O(n^2)\) 时间。假设我们按照图 12-2 所示的方式，将数组从中点分为两个子数组，则划分需要 \(O(n)\) 时间，排序每个子数组需要 \(O((n / 2)^2)\) 时间，合并两个子数组需要 \(O(n)\) 时间，总体时间复杂度为：

\[ O(n + (\frac{n}{2})^2 \times 2 + n) = O(\frac{n^2}{2} + 2n) \]

----这里推导公式是不是多了写特殊符号？markdown解析问题吗

**Reply by krahets** [reply]
应该是网络问题，试试换个网络环境，比如手机 4G

**Comment by clearwater753** [comment]
作者您好，我想问一下为什么将数组从中点分为两个子数组，划分需要O(n)时间？
直接计算中点不是O(1)时间吗？中点计算好了也就划分好了？

**Reply by lrisora** [reply]
最后一步排序结果需要存入原位，所以这里的子数组是新申请的内存，需要整体复制
如果直接将原内存二分原地冒泡的话
接下来的合并就需要将这8个数据先整体复制出来
才能在原内存生成合并结果，这使得合并复杂度变为O(2n)
整体是一致的

**Comment by Aidener** [comment]
这个证明漂亮，比起一些支支吾吾半天讲不出所以然的舒服多了

**Comment by qw124341789** [comment]
归并排序真的是太“分治”了

**Comment by bnufw** [comment]
以“冒泡排序”为例......
这是笔误了嘛，应该是以"归并排序"为例吧？

**Reply by bnufw** [reply]
嗷嗷，懂了

**Comment by bnufw** [comment]
在1.   操作数量优化 的例子中
划分的复杂度为什么是O(n)？我的想法是，由于len()函数的复杂度为O(1),那就直接len/2
对么？

**Reply by krahets** [reply]
Hi，这里的划分是指划分为两个子数组，涉及遍历 + 赋值操作，因此时间复杂度为 O(n)

**Reply by 480284856** [reply]
hello, K神，我觉得“分”阶段的把一个原数组划分为两个子数组的过程，应该是计算middle的过程吧，这里是不涉及遍历和赋值操作的，只有在“治”的阶段，才会遍历两个已排好序的子数组，并进行“两个子数组的值排序好，放到tmp数组中"以及"把tmp数组中的值放回nums"的赋值操作吧。

```python
def merge_sort(nums: list[int], left: int, right: int):
    if left==right:
        return
    
    # 把一个长数组划分为两个子数组，分别排好序
    mid=int((left+right)/2)
    merge_sort(nums, left, med)
    merge_sort(nums, med+1, right)

    merge(nums, left, mid, right)
```

**Reply by 480284856** [reply]
所以我感觉分治一次的冒泡排序的时间复杂度应该是 $O(2(\frac{n}{2})^2 + n)$

**Reply by 480284856** [reply]
我明白了，冒泡排序函数的参数只有一个nums，所以得复制出来。

**Comment by Kore-yoshi** [comment]
我我我有一个很大的疑问，可以说递归这个思想理解起来还是比较容易，但是如何使用它，怎么能够想起来使用它好像是一个很大的问题哎。
就以leetcode的21题合并两个有序链表为例子，题目：将两个升序链表合并为一个新的 升序 链表并返回。新链表是通过拼接给定的两个链表的所有节点组成的。
这个题是如何想起来能够使用递归的？希望各位佬能够耐心解释一下（根本想不出来，不知道是不是自己对递归的理解有误或者是不太深）

**Reply by hpstory** [reply]
Hi @Kore-yoshi , [看看这篇](https://oi.wiki/basic/divide-and-conquer/)，里面的例子可以加深你对递归的理解。递归的基本思想是某个函数直接或者间接地调用自身，这样原问题的求解就转换为了许多性质相同但是规模更小的子问题。

**Reply by krahets** [reply]
递归的关键在于定义“问题及子问题之间的关系”。例如将“将两个升序链表合并为一个新的升序链表”。那么假设两个链表的长度分别为 m, n，设原问题为 f(m, n)。假设我们先合并了第一个链表的首节点，那么剩余的就是子问题 f(m-1, n) ；以此递归下去，最终达到 f(0, 0) 时问题就能得到解决。

**Reply by fuochai** [reply]
递归的关键在于 子问题不需要上一层的信息
有一个任务 你做了一步之后把剩下的外包给别人 他不需要知道你做过的那部分的是怎样的
这个问题你不说我还真没想到能递归 直接做也没啥问题 因为它是一分一的
有些时候递归明显更方便 像二叉树中序遍历 是一分二的 中间还要加一个同级的结果

**Reply by z1y2y3** [reply]
我和你一样的感受，现在我大致懂了。递归这个思想非常深刻（递归函数和图灵机等价，任何编程语言（任何程序）任何计算机都是计算能力上等价的），首先递归和迭代（for循环，while循环）的计算能力等价，也就是while循环写出的程序可以用递归写出来，反过来也可以，这么说，禁用for和while，我们仍然可以全部用递归写出题解来，反过来也一样，甚至可以把for和while循环看作是递归的语法糖，所以“如何想起来用递归”这个问题不是个问题，你完全可以不用循环只用递归写程序，只不过你习惯了用循环。所以多用递归解题可以帮助你。

**Reply by z1y2y3** [reply]
怎么写递归才是关键，怎么非常自然的写递归或者说写暴力搜索。任何递归都是dfs，任何dfs都是在一个状态空间上遍历，任何问题都能dfs解决所以任何问题都是在某个状态空间上的遍历（前提是这个问题是可计算的，当然lc上的算法题都是可计算的，我想大部分工程问题也是可计算的）。明确了三点你就可以写出dfs：状态定义，状态转移，基本状态（base case）。根本上是找状态定义（一般有了状态推理就得到状态转移（做多了就会觉得不难），base case是很容易找），但是怎么找呢，大部分教材或者题解都是说“你可以先定义这样一个状态”“你可以假设xxx”，这实际上是把三点直接给你了，并没有说这状态是怎么找到的。所以，之前说的都是废话，但都是正确的而且必要的。

**Reply by z1y2y3** [reply]
怎么写dfs呢？
首先纠正一点，找状态和找状态转移有时是需要同时进行的，并不一定是先找状态，再找状态转移。
然后是有了三点，怎么写dfs，这是模板：
```cpp
void dfs(State s) {
    if (isTarget(s)) return;
    for (State next : expand(s)) {
        dfs(next);
    }
}
```
上面的next状态改成before状态也可以用来构建当前状态，expand改成s.getAllNeighbors()可能更好，这个模板改改就能根据三点写出代码了，学过图的dfs和bfs就更好理解，把State看成节点。
最后是怎么找状态和状态转移，也就是中间状态法（有时叫做子问题法可能更自然），拿到任何一道题我们可以考虑暴力搜索，脑子里的过程是：我在“走” → 我看到一个中间局面 → 用数据结构记录和描述这个局面。1我在“走”，是你对于这道题的某一个思路，这个思路上所形成的所有中间状态，你在中间状态组成的状态空间上转移，从中间状态考虑，去总结和描述任意时刻的某个局面是关键，当前前提是你得有一个暴力的思路，比如生成一个集合的所有子集，思路就是n个元素每个元素选或不选，面临的一个局面就是我已经选好前i个元素了，在考虑当前第i+1个元素选或不选（状态和状态转移同时出现了）。2记录局面就是用全局变量、dfs函数参数和返回值、dfs函数局部变量来表达描述这个局面所需要的全部信息，3状态（state）就是描述“当前问题局面”所需的全部信息，一个中间局面就是面对的一个中间状态。  可以有更多例子（所有的题目都可以说例子。。。），现在已经足够写出一个问题的暴力搜索了。
一些常识:  1写dfs的时候要“相信自己的递归是对的”，明确递归函数的定义，相信你的递归能做到。也就是说你面对的一个局面，总是要相信有些东西是已经做好了的，因为毕竟是一个中间局面，怎么得到的就要假设我以某种方式到了这个局面 。2你一进入dfs的函数体，实际上就拿到了当前这个中间状态的所有信息，你就可以做一些事情，比如用这个中间状态更新某种max值，比如记录一些什么信息等等。

**Comment by xiaozhuge123** [comment]
day10
